torch/csrc/jit/argument_spec.h

```diff
@@ -32,34 +34,38 @@ struct TensorInfoPOD {
                        // for tensor 0, the offset is always 0
 };
 
-static_assert(sizeof(TensorInfoPOD) == sizeof(int64_t),
-  "TensorInfoPOD must be 64-bit struct for ArgumentSpec encoding to work");
+static_assert(sizeof(ArgumentInfoPOD) == sizeof(int64_t),
+  "ArgumentInfoPOD must be 64-bit struct for ArgumentSpec encoding to work");
 
-struct TensorInfo;
+struct ArgumentInfo;
 
 struct ArgumentSpec {
-  // note: tensors must always be variables
-  ArgumentSpec(bool with_grad, const variable_tensor_list & tensors)
-  :  hash_code(0), ntensors(tensors.size()) {
-    int all_dims = 0;
-    for(size_t i = 0; i < ntensors; i++) {
-      all_dims += tensors[i].defined() ? tensors[i].ndimension() : 0;
+  ArgumentSpec(bool with_grad, const Stack & inputs)
+  :  hash_code(0), ninputs(inputs.size()) {
+    int32_t all_dims = 0;
+    const int32_t num_inputs = inputs.size();
+    for (int32_t i = 0; i < num_inputs; i++) {
+      if (!inputs[i].isTensor()) continue;
+      auto tensor = inputs[i].toTensor();
+      all_dims += tensor.defined() ? tensor.ndimension() : 0;
     }
     // allocate enough room for all TensorPODs and dimensions
-    data.resize(ntensors + all_dims*2);
+    data.resize(ninputs + all_dims*2);
 
     // and reinterpret our data array as these structs
-    TensorInfoPOD * pods = reinterpret_cast<TensorInfoPOD*>(data.data());
+    ArgumentInfoPOD * pods = reinterpret_cast<ArgumentInfoPOD*>(data.data());
     int64_t * next_dim = sizes_strides();
-    int total_dims = 0;
-    for(size_t i = 0; i < ntensors; i++) {
-      const auto & t = tensors[i];
+    int32_t total_dims = 0;
+    for(int32_t i = 0; i < num_inputs; i++) {
       auto & pod = pods[i];
+      pod.kind = static_cast<uint32_t>(inputs[i].kind());
+      if (!inputs[i].isTensor()) continue;
+      const auto & t = inputs[i].toTensor();
```

Comment:
`toTensor()` returns by value, so this is a reference to a temporary that dies at the end of the line, right?

Reply:
Yep, good point